# Overview

The goal of this lab is to familiarize yourself with standard data wrangling tasks using Python. Being able to do so will greatly facilitate any work you do from here onward. After this lab, you should be comfortable enough in a Python data environment that we can begin using a variety of packages to analyze and visualize different kinds of data more freely.

## Your second (and last) week as a data specialist at Via Rail

Your supervisor at Via Rail has handed over more 'information'. This time, it's a few poorly formatted tables stored as *.txt* file. They want a basic map, and since Via is *finally* trying to enter the digital age with all this new federal funding, they want this map to create itself with the click of a button (in this case, by running all code cells in this notebook). Like last week, you assume there will be more data coming your way in this format, so you will need to come up with an automated process that imports it, parses it, and does any subesquent operations that are necessary in order to finally prep it for the mapping intern. In other words, you need to write a program that takes data of a certain format as input, and generates clean data as an output.

## Importing packages

At the start of any Python script (or Jupyter notebook running Python) you will usually want to import packages so that their functionality is available to the rest of your code. One of Python's great strengths is the quality of its 3rd party packages, which can provide functionality for doing anything from video editing to running a web server to... managing data! In this session, we would like to import the [Pandas](https://pandas.pydata.org/) package. Pandas allows us to manipulate data stored in Pandas dataframes, and forms the bedrock of most standard data analysis done with Python today. Familiarity with other basic data management software (e.g. Excel) will serve you well here.

In [120]:
#the below command imports pandas
import pandas

#when importing a package, you can give it the alias of your choice by appending 'as' followed by your alias.
#doing so can make it more practical for you to refer to the package in your code, especially if the package were to have a really long name...
import pandas as pd

#from here on, we can refer to pandas simply as 'pd'

## Inspecting and importing data

Make sure you *know* your data. The first thing you should always do before importing data is to inspect it.
- From your command line or file explorer (Windows Explorer/Finder), open *stations.txt* in a text editor (Notepad/Text Edit)

What do you observe? Look for patterns in the data: can you spot any characteristics that you could use to parse it into a table? Ideally, you would like to convert this into a clean table where each data point is separated into its own cell under a header. If you are good in Excel, you could easily clean this up in that software. However, you want your process to be as automated as possible, so your initial script will involve some of what you did last week.

1. What would you say is the 'delimiter' or 'separator' for this data? Doesthe data have headers? Indicate your answers in a Markdown cell below.

Each entry is delimited with semi-colons (;), which would correspond to different variables separated into columns in a dataframe. However, these variables/columns do not have names/headers.

- close the file *stations.txt*.

Now you're going to import this dataset. Pandas has a method called [*read_csv*](https://Pandas.pydata.org/docs/user_guide/io.html#csv-text-files), which can be used to import almost any kind of table into a [Pandas dataframe](https://pandas.pydata.org/docs/user_guide/dsintro.html#dataframe). Although your data doesn't have headers, there is a file in this lab which lists these...

The *read_csv* Pandas method can take a series of optional parameters, which we call 'arguments' and 'keyword arguments' respectively. **To continue, you will need to indicate a value for each of the two following variables.**


In [121]:
#we will begin by filling a pathname so the read_csv method can find the stations.txt file
#you may need to modify this pathname to reflect the location of the stations.txt file on your computer
path_stations = "Lab 3 Data Wrangling Stations.txt"

#a string (my_sep) will be passed to the 'sep' argument of the read_csv method
#and the other, a list of strings (my_headers), will be used by the 'names' argument. You can find these headers in headers.txt
my_sep = ";"
my_headers = "object_ID;stat_num;stat_ID;stat_name;street_num;street;address_2;city;province;postal_code;country" #copy this in from the headers.txt file
my_headers = my_headers.split(my_sep)

df_stations = pd.read_csv(path_stations, sep = my_sep, header = None, names = my_headers)

Yay! Now you can start working with Pandas.

You now have a Pandas dataframe stored in a variable called *df_stations*.
The print statement can be used to display the dataframe once imported, but Jupyter is able to render dataframes in a more appealing way. Find active variables in the side panes of Google Colab, or, in VS Code, under the 'Jupyter' tab of the bottom tray (beside the terminal).

In [122]:
print(df_stations)

     object_ID  stat_num stat_ID                                 stat_name  \
0            1       555    XYUL  Aeroport Montreal Pierre-Elliott Trudeau   
1            2       600    ALDR                                 Aldershot   
2            3       344    ALEX                                Alexandria   
3            4       221    AMQU                                     Amqui   
4            5       636    AMYO                                     Amyot   
..         ...       ...     ...                                       ...   
204        205       299    WNDG                                   Windigo   
205        206       618    WDON                                   Windsor   
206        207       128    WOMR                               Woman River   
207        208       282    WDST                                 Woodstock   
208        209       280    WYOM                                   Wyoming   

     street_num                                 street  address

## Exploring Pandas dataframes

You can see that not all your data is visible. It would take up a lot of space to display all your data like an Excel sheet (besides, you have other software for doing that). You can use the following dataframe methods to explore your data in Pandas.

In [123]:
#head will show you the top 5 rows
df_stations.head()

#you can also add a number indicating how many rows you would like to see inside the parentheses ()

object_ID  stat_num stat_ID                                 stat_name  \
0          1       555    XYUL  Aeroport Montreal Pierre-Elliott Trudeau   
1          2       600    ALDR                                 Aldershot   
2          3       344    ALEX                                Alexandria   
3          4       221    AMQU                                     Amqui   
4          5       636    AMYO                                     Amyot   

   street_num                                 street  address_2  \
0         NaN  Boulevard Roméo-Vachon Nord (départs)        NaN   
1         NaN                            East Tunnel        NaN   
2        45.0                  McDougald Street East        NaN   
3         NaN           Boulevard Saint-Benoît Ouest        NaN   
4         NaN                            Road 8 & 20        NaN   

                       city province postal_code country  
0                       NaN   Quebec     H4Y 0A4  Canada  
1                Burlington  Ontario     L7T 2C4  Canada  
2           North Glengarry  Ontario     K0C 1A0  Canada  
3                       NaN   Quebec     G5J 2E8  Canada  
4  Unorganized North Algoma  Ontario         NaN  Canada

You can also use the tail method to show the last rows of a dataframe.

2. Use the tail method with the appropriate argument in the code cell below to show the last 15 rows of *stations_df*.

In [124]:
print(df_stations.tail(15))

     object_ID  stat_num stat_ID         stat_name  street_num  \
194        195       119    TRTO           Toronto       140.0   
195        196       413    TRNJ  Trenton Junction         NaN   
196        197       170    TPIS    Trois Pistoles         NaN   
197        198       247    VBRU     Van Bruyssels         NaN   
198        199       339    VDRY            Vandry         NaN   
199        200        59    WSHG           Washago         NaN   
200        201        76    WSTR           Westree         NaN   
201        202        49    WYMT           Weymont         NaN   
202        203       116    WHTR       White River         NaN   
203        204       168    WGWC            Wigwam         NaN   
204        205       299    WNDG           Windigo         NaN   
205        206       618    WDON           Windsor         NaN   
206        207       128    WOMR       Woman River         NaN   
207        208       282    WDST         Woodstock       100.0   
208       

Below are some other commands you can use to explore your data...

In [125]:
#shape will return a tuple containing the number of rows followed by the number of columns.
df_stations.shape

(209, 11)

In [126]:
#info will show you useful information related to your data
df_stations.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 209 entries, 0 to 208
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   object_ID    209 non-null    int64  
 1   stat_num     209 non-null    int64  
 2   stat_ID      209 non-null    object 
 3   stat_name    209 non-null    object 
 4   street_num   28 non-null     float64
 5   street       161 non-null    object 
 6   address_2    0 non-null      float64
 7   city         100 non-null    object 
 8   province     209 non-null    object 
 9   postal_code  99 non-null     object 
 10  country      209 non-null    object 
dtypes: float64(2), int64(2), object(7)
memory usage: 18.1+ KB


In [127]:
#describe will compute some summary statistics about your data
#note that such statistics may be meaningless depending on the nature of your data
df_stations.describe()

object_ID    stat_num   street_num  address_2
count  209.000000  209.000000    28.000000        0.0
mean   105.000000  250.488038   382.964286        NaN
std     60.477268  176.179190   738.454215        NaN
min      1.000000    1.000000     1.000000        NaN
25%     53.000000   95.000000    48.750000        NaN
50%    105.000000  217.000000   166.000000        NaN
75%    157.000000  390.000000   365.000000        NaN
max    209.000000  639.000000  3875.000000        NaN

3. Based on the information provided by the commands above, how many stations are there in your data?

4. Based on the information provided by the commands above, how many different cities are represented in your data? You may wish to run another command to better elucidate this info.

In [128]:
print(df_stations.shape)

(209, 11)


From the shape function we can see there are 209 stations

5. Why do you think the *describe()* method is only providing summary stats for a limited number of fields?

The describe fuction produces statistical stats of each column. It can only describe columns that have numeric values as you cannot find the mean of a column full of strings.

You can easily create new dataframes by subsetting; or, in other words, selecting specific columns to assign to a new dataframe. This is great for shedding columns you might not need. Subsetting a dataframe has similarities to how you would do it with a dictionary: You use the [] to select fields by their header names. See below:

In [129]:
#to isolate a single column, simply select the header like you would a key in a dictionary.
#This creates a Pandas 'series', which is kind of analogous to a list. It's basically a one-dimensional dataframe...
station_names = df_stations['stat_name']
station_names.head()

#pass station_names to the type() function. What data type is it?
type(station_names)

pandas.core.series.Series

In [130]:
#for example, let's create a new dataframe that only contains the station ID and station name.
#to subset multiple columns to a new dataframe, you will need to pass a list of these headers
little_df = df_stations[['stat_ID','stat_name']]
little_df.head()

stat_ID                                 stat_name
0    XYUL  Aeroport Montreal Pierre-Elliott Trudeau
1    ALDR                                 Aldershot
2    ALEX                                Alexandria
3    AMQU                                     Amqui
4    AMYO                                     Amyot

6. In the code cell below, create a new dataframe called *temp_df* that only contains the following fields: *stat_ID*, *stat_name*, *city*, *postal_code*

In [131]:
temp_df = df_stations[["stat_ID", "stat_name", "city", "postal_code"]]

print(temp_df.head())

  stat_ID                                 stat_name                      city  \
0    XYUL  Aeroport Montreal Pierre-Elliott Trudeau                       NaN   
1    ALDR                                 Aldershot                Burlington   
2    ALEX                                Alexandria           North Glengarry   
3    AMQU                                     Amqui                       NaN   
4    AMYO                                     Amyot  Unorganized North Algoma   

  postal_code  
0     H4Y 0A4  
1     L7T 2C4  
2     K0C 1A0  
3     G5J 2E8  
4         NaN  


Another way of shedding unneeded columns is simply to delete them in-place (within the original dataframe). In *df_stations*, There are columns which you could do without. These would be the following: *object_ID*, *address_2*

These two fields are by most measures useless. You can tell because they are either zombie ID fields, or contain so few values that their use in any scenario is questionable. Search for the delete method in the [Pandas documentation](https://pandas.pydata.org/docs/user_guide/index.html) and delete these fields!

(hint: you might want to check inside the page on dataframes...)

Once you find the relevant information, you will see that there are two common ways of deleting fields: 'del', and the more familiar 'pop' method used with lists.

7. Use the pop method to delete these fields.

In [132]:
#use the pop method


df_addy = df_stations.pop('address_2')
df_obj = df_stations.pop('object_ID')



# # #then run the cell to display your dataframe below and make sure everything is as expected
# df_stations.head()
print(df_stations)

     stat_num stat_ID                                 stat_name  street_num  \
0         555    XYUL  Aeroport Montreal Pierre-Elliott Trudeau         NaN   
1         600    ALDR                                 Aldershot         NaN   
2         344    ALEX                                Alexandria        45.0   
3         221    AMQU                                     Amqui         NaN   
4         636    AMYO                                     Amyot         NaN   
..        ...     ...                                       ...         ...   
204       299    WNDG                                   Windigo         NaN   
205       618    WDON                                   Windsor         NaN   
206       128    WOMR                               Woman River         NaN   
207       282    WDST                                 Woodstock       100.0   
208       280    WYOM                                   Wyoming       579.0   

                                    street         

8. Once you have run your deletions in the cell above, run the code cell again. Why does it fail? What is a KeyError?

It fails becuase we have already deleted the columns. So when we go back to delete it again the computer cant find the columns (they've already been deleted) giving us a key error.

## Preparing your data for the mapping intern

Now that your table is all cleaned up, it's time to prepare it for mapping. There's one issue though: the data you've been working with so far doesn't seem to have any explicit location information (i.e. coordinates). There is other geographic information in your table, however, which you could potentially exploit for mapping purposes. Can you name which fields these would be?

9. Which fields do you think you would need in order to [geocode](https://desktop.arcgis.com/en/arcmap/latest/manage-data/geocoding/what-is-geocoding.htm) these transit stations? Name all that apply in the Markdown cell below.

You could use the street name in combination with the street number although these fields are missing in some of the rows. You could also use postal code or most likely a combination of all 3.

Luckily, you won't need to be doing any geocoding today, since you noticed that your supervisor also left you with another file called *coordinates.txt*.

10. Open the file in a text editor and inspect it like you did with the stations file. Do you notice a pattern here which you could use for parsing the data?
- What would the delimiter be for this data? What file extension do we typically use to associate to this delimiter?
- Does the data have headers?
- Can you identify a field in *coordinates.txt* that you could use to merge with your stations data? Which field?

The delimeters are ',' and normally this file extension would be '.csv'. Yes the data does have headers: lat,lon,stat_ID. We could use the stat_ID to merge the two datasets.


11. Import *coords.txt* into JupyterLab using the appropriate Pandas method (you've done this before!) and assign the dataframe to a variable called *df_coords*.

In [133]:
path_coords = "Lab 3 Coordinates.txt"

df_coords = pd.read_csv(path_coords)
print(df_coords)

           lat        lon stat_ID
0    45.456988 -73.751834    XYUL
1    43.312886 -79.855009    ALDR
2    45.318050 -74.639672    ALEX
3    48.466260 -67.436172    AMQU
4    48.482877 -84.954122    AMYO
..         ...        ...     ...
204  47.771220 -73.334097    WNDG
205  42.324933 -83.007134    WDON
206  47.511624 -82.628067    WOMR
207  43.126528 -80.752123    WDST
208  42.947960 -82.122260    WYOM

[209 rows x 3 columns]


## Merging dataframes

You will need to join your coordinate data with your station information. You can do this using the Pandas [merge method](https://pandas.pydata.org/docs/user_guide/merging.html#database-style-dataframe-or-named-series-joining-merging).

Before doing a merge, you should identify which column can be used to perform the merge. [Merges](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.merge.html) are very similar to standard [join operations](https://dataschool.com/assets/images/how-to-teach-people-sql/sqlJoins/sqlJoins_7.png) in SQL, which you have probably seen in software like ArcGIS before (e.g. joining some data to the attribute table of a shapefile based on a common identifier).

To merge tables, you will need to identify a column that contains values which can be used to match rows from each table to eachother.

**Take note of which column you could use as a join field for merging *df_stations* with *df_coords*.**

Before performing a merge, you want to make sure there are no duplicates in your data. A duplicate could mean that the same row from the table being merged gets joined twice or more. The [duplicated method](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.duplicated.html) can serve you well here.

In [134]:
duplicates = df_coords.duplicated()
print(type(duplicates))
#below is a little loop for verifying if there are any duplicates (note that there are simpler ways of doing this with Pandas than using a for loop)
for x in duplicates:
    if x == True:
        print(x)
    else:
        pass

<class 'pandas.core.series.Series'>


12. What data type is being generated by the duplicated() method? Are there any duplicates in *df_stations*? Run a code cell to determine this, then a short text explanation in a Markdown cell.

In [135]:
if duplicates.any(): # I used Ai to find the .any fuction, i tried to use "if True in duplicates" but that didnt work with the list so i needed to find another way 
    print("There are duplicate rows")
else:
    print("There are no duplicate rows!!")

There are no duplicate rows!!


The duplicates method is generating a pandas series of boolean true of false values, if a row has a duplicate its cooreponding position in the duplicates series would display true. There are no duplicates. In the code block above, I am checking if "True" is wihtin the entire list. If yes this means there is at least 1 duplicate row, if no this means all the values are false and there are no duplicates. 

The code below will test whether there are any duplicates specifically within the *stat_ID* field, which could be a good field for eventually merging the data...

In [136]:
duplicates = df_stations['stat_ID'].duplicated()

for x in duplicates:
    if x == True:
        print(x)
    else:
        pass

### Using a custom function to check for duplicates

[Functions](https://www.w3schools.com/python/python_functions.asp) are a great way to organize your code. Basically, you can wrap any code inside a function, which allows you to call that code from anywhere else in your code without having to rewrite the code all over again! Think of it as a more elaborate variable, but a variable that *does* stuff, instead of just holding static information. Like variables, you need to declare a function first before using it. After that, you can refer to it anywhere, as long as you pass it the right arguments!

Functions in Python are declared using the 'def' keyword. Immediately following the name of the function - which is up to you to choose - and much like any variable name, you need to add parentheses (()). Inside these parentheses is where you can pass arguments to the function (i.e. input data).

Like conditional statements or loops, function declarations are immediately followed by a colon (:), and any code nested inside a function is indented. The 'return' statement at the end of a function signals what it will output. Think of the () of the function as the point of entry, and the 'return' statement as the exit.

13. You would like to write a custom function to check for duplicates so that you can freely call it from anywhere in your code. Your function should require a **list of booleans as input**, and **return a number indicating the number of duplicates** (i.e. True values) in the list that was passed to it. Use the documentation provided above to familiarize yourself with functions (and any other documentation you can find).

Hint: len() will provide the length of a list (and therefore a number). .append() can be used for appending True values to a list, which you can then use len() on as your return value.

In [137]:
def dup(duplicates_list):
    if duplicates_list.any():
        print("There are duplicates: ")
        num_trues = 0
        for x in duplicates_list:
            if x == True:
                num_trues = num_trues + 1
            else:
                pass
        print("There are", num_trues, "duplicate rows")
    else:
        print("There are no duplicate rows!!")

14. How many duplicates are there in the *city* field of df_stations? Produce the result in the code below.

In [138]:
city_dups = df_stations['city'].duplicated()

dup(city_dups)

# df_stations.to_csv('df_stations.csv',header=True,index=True)
# city_dups.to_csv('city_dups.csv',header=True,index=True)   

There are duplicates: 
There are 163 duplicate rows


### Merging at last!

Now that you are certain there are no duplicates, you can feel confident about merging these tables. 

15. Finally, use the documentation (links) provided above (under the **Merging dataframes** title) to perform this merge. You can also use Google or ask an AI chatbot if you need other resources. You will need to do your own research here. However, the task shouldn't be too complex, and you shouldn't have to use too many of the optional parameters (keyword arguments) that the merge method provides. Assign your merged data to a dataframe called *df_stations_coords*.

In [139]:
df_merged = df_stations.merge(df_coords,on='stat_ID',how='inner')
print(df_merged)
df_merged.to_csv('print(df_merged).csv',header=True,index=True)   

     stat_num stat_ID                                 stat_name  street_num  \
0         555    XYUL  Aeroport Montreal Pierre-Elliott Trudeau         NaN   
1         600    ALDR                                 Aldershot         NaN   
2         344    ALEX                                Alexandria        45.0   
3         221    AMQU                                     Amqui         NaN   
4         636    AMYO                                     Amyot         NaN   
..        ...     ...                                       ...         ...   
204       299    WNDG                                   Windigo         NaN   
205       618    WDON                                   Windsor         NaN   
206       128    WOMR                               Woman River         NaN   
207       282    WDST                                 Woodstock       100.0   
208       280    WYOM                                   Wyoming       579.0   

                                    street         

Once you are finished, from the toolbar at the top of JupyterLab/Colab, click Kernel > **Restart Kernel and Clear All Outputs...**. Once the notebook has completed rebooting, click Run > **Run All Cells**. Make sure that all your code runs properly, and produces your merged table at the end.

---

Fed up with being an intermediary between your supervisor's lousy data and the pesky mapping interns, you decide to take on greater challenges. You can do the mapping, AND you can find quality data for doing so. Further, you have ambitions: you want to explore more interesting data that you can analyze and do research with.

# Deliverables

Complete the questions in this notebook and make sure all your code runs, then upload the notebook file (*.ipynb*) to the Assignment page on Moodle. If you ran this lab in Colab, you will need to export/download the notebook file in order to upload it.

Optionally, you can also create another GitHub repository for this lab and push/commit this lab file to it. For the upcoming labs, no use of GitHub repos will be necessary for marks, but they will come in handy towards the end of the term for the hosting of your webmap project, so familiarising yourself with GitHub may not be the worst idea...